File: android/src/main/java/com/example/haptic_composer/HapticComposerPlugin.java
package com.example.haptic_composer;

import android.content.Context;
import android.os.Build;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.os.VibratorManager;
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodCall;

public class HapticComposerPlugin implements FlutterPlugin {
    private static final String CHANNEL = "com.example.haptic_composer/haptic";
    private MethodChannel channel;
    private Context context;
    private Vibrator vibrator;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        context = binding.getApplicationContext();
        channel = new MethodChannel(binding.getBinaryMessenger(), CHANNEL);
        channel.setMethodCallHandler(((call, result) -> {
            switch (call.method) {
                case "initialize":
                    initialize(result);
                    break;
                case "triggerEffect":
                    triggerEffect(call, result);
                    break;
                case "isSupported":
                    isSupported(result);
                    break;
                case "dispose":
                    dispose(result);
                    break;
                default:
                    result.notImplemented();
            }
        }));
        initializeVibrator();
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private void initialize(MethodChannel.Result result) {
        try {
            initializeVibrator();
            result.success(vibrator != null && vibrator.hasVibrator());
        } catch (Exception e) {
            result.success(false);
        }
    }

    private void initializeVibrator() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            VibratorManager vibratorManager = (VibratorManager) context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE);
            vibrator = vibratorManager != null ? vibratorManager.getDefaultVibrator() : null;
        } else {
            vibrator = (Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE);
        }
    }

    private void triggerEffect(MethodCall call, MethodChannel.Result result) {
        Double intensity = call.argument("intensity");
        Integer duration = call.argument("duration");

        if (intensity == null) intensity = 0.5;
        if (duration == null) duration = 100;

        try {
            if (vibrator == null) {
                initializeVibrator();
            }

            if (vibrator != null && vibrator.hasVibrator()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    int amplitude = (int) (intensity * 255);
                    amplitude = Math.max(0, Math.min(255, amplitude));
                    VibrationEffect effect = VibrationEffect.createOneShot(duration.longValue(), amplitude);
                    vibrator.vibrate(effect);
                } else {
                    vibrator.vibrate(duration.longValue());
                }
                result.success(null);
            } else {
                result.error("NO_VIBRATOR", "Device does not support vibration", null);
            }
        } catch (Exception e) {
            result.error("VIBRATION_ERROR", "Failed to trigger vibration: " + e.getMessage(), null);
        }
    }

    private void isSupported(MethodChannel.Result result) {
        try {
            boolean supported = vibrator != null && vibrator.hasVibrator();
            result.success(supported);
        } catch (Exception e) {
            result.success(false);
        }
    }

    private void dispose(MethodChannel.Result result) {
        try {
            if (vibrator != null) {
                vibrator.cancel();
            }
            result.success(null);
        } catch (Exception e) {
            result.error("DISPOSE_ERROR", "Failed to dispose: " + e.getMessage(), null);
        }
    }
}

File: android/app/src/main/kotlin/com/example/haptic_composer/HapticComposerPlugin.kt
package com.example.haptic_composer

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import androidx.annotation.RequiresApi
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel

class HapticComposerPlugin: FlutterPlugin {
    private val CHANNEL = "com.example.haptic_composer/haptic"
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var vibrator: Vibrator? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "initialize" -> initialize(result)
                "triggerEffect" -> triggerEffect(call, result)
                "isSupported" -> isSupported(result)
                "dispose" -> dispose(result)
                else -> result.notImplemented()
            }
        }
        initializeVibrator()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun initialize(result: MethodChannel.Result) {
        try {
            initializeVibrator()
            result.success(vibrator != null && vibrator!!.hasVibrator())
        } catch (e: Exception) {
            result.success(false)
        }
    }

    private fun initializeVibrator() {
        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as? VibratorManager
            vibratorManager?.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
        }
    }

    private fun triggerEffect(call: MethodChannel.MethodCall, result: MethodChannel.Result) {
        val intensity = call.argument<Double>("intensity") ?: 0.5
        val duration = call.argument<Int>("duration") ?: 100

        try {
            if (vibrator == null) {
                initializeVibrator()
            }

            if (vibrator?.hasVibrator() == true) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    // Use VibrationEffect for Android O and above
                    val amplitude = (intensity * 255).toInt().coerceIn(0, 255)
                    val effect = VibrationEffect.createOneShot(duration.toLong(), amplitude)
                    vibrator?.vibrate(effect)
                } else {
                    // Fallback for older Android versions
                    @Suppress("DEPRECATION")
                    vibrator?.vibrate(duration.toLong())
                }
                result.success(null)
            } else {
                result.error("NO_VIBRATOR", "Device does not support vibration", null)
            }
        } catch (e: Exception) {
            result.error("VIBRATION_ERROR", "Failed to trigger vibration: ${e.message}", null)
        }
    }

    private fun isSupported(result: MethodChannel.Result) {
        try {
            val supported = vibrator?.hasVibrator() ?: false
            result.success(supported)
        } catch (e: Exception) {
            result.success(false)
        }
    }

    private fun dispose(result: MethodChannel.Result) {
        try {
            vibrator?.cancel()
            result.success(null)
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", "Failed to dispose: ${e.message}", null)
        }
    }
}
